Bluetooth: Controller: Use HAL_TICKER_US_TO_TICKS_CEIL for ticks_slot

cvinayak opened the pull request:
Use HAL_TICKER_US_TO_TICKS_CEIL for ticks_slot calculations.

cvinayak committed:
Bluetooth: Controller: Use HAL_TICKER_US_TO_TICKS_CEIL for ticks_slot

Use HAL_TICKER_US_TO_TICKS_CEIL for ticks_slot calculations.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Revert EVENT_OVERHEAD_END_US to original value

Revert back the EVENT_OVERHEAD_END_US value to original
value in initial commit 5c3709c13fd8 ("bluetooth:
controller: Add def's re. update to slot reservation calc").

40 us is a sufficient margin considering ~30 us being the
worst case Rx chain delay for S8 Coded PHY on nRF5x SoCs.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Make openisa and test hal ticker.h consistent

Make openisa and unit test hal ticker.h consistent with
nrf hal ticker.h.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

PavelVPV commented on tests/bsim/bluetooth/mesh/_mesh_test.sh:
```diff
@@ -68,8 +70,9 @@ function RunTest(){
-      -testid=$testid ${testids["${testid}"]} ${test_options}
+      -rs=$seed -testid=$testid ${testids["${testid}"]} ${test_options}
     let idx=idx+1
+    let seed=seed+1
```
Could you elaborate why this change is needed? Will mesh tests work without it?

cvinayak replied:
pb_remote_reprovision.sh   was failing in this PR.. Now I see that other tests are failing when devices have different random seed values.

Debuging with @m-alperen-sener, we were able to pass  pb_remote_reprovision.sh by having a small (2ms) delay in relay_send.

Instead of adding delay, we decided to have random seed value in test script

PavelVPV replied:
We used to add/increase delays in a particular test in case when the issue is not caused by invalid implementation (for example: https://github.com/zephyrproject-rtos/zephyr/commit/e5fcca6e99a52334c0c726dfc1b0c432f8d944bb#diff-7df14fc51b85bcb032356d6314ea93067acb2b55ad69b8585edd8821709da43c). Changing the seed will affect other tests.

PavelVPV replied:
I can have a look later today

PavelVPV replied:
Please try this patch:
```
diff --git a/tests/bsim/bluetooth/mesh/src/test_provision.c b/tests/bsim/bluetooth/mesh/src/test_provision.c
index 171f8cccbb..36e3c1250a 100644
--- a/tests/bsim/bluetooth/mesh/src/test_provision.c
+++ b/tests/bsim/bluetooth/mesh/src/test_provision.c
@@ -554,26 +554,26 @@ static void node_configure_and_reset(void)
                                          &status));
        ASSERT_EQUAL(0, status);
 
-       k_sleep(K_SECONDS(1));
+       k_sleep(K_SECONDS(2));
 
        ASSERT_OK(bt_mesh_cfg_cli_mod_app_bind(0, current_dev_addr, current_dev_addr, 0x0,
                                           BT_MESH_MODEL_ID_HEALTH_SRV, &status));
        ASSERT_EQUAL(0, status);
 
-       k_sleep(K_SECONDS(1));
+       k_sleep(K_SECONDS(2));
 
        ASSERT_OK(bt_mesh_cfg_cli_mod_sub_add(0, current_dev_addr, current_dev_addr, 0xc000,
                                          BT_MESH_MODEL_ID_HEALTH_SRV, &status));
        ASSERT_EQUAL(0, status);
 
-       k_sleep(K_SECONDS(1));
+       k_sleep(K_SECONDS(2));
 
        ASSERT_OK(bt_mesh_cfg_cli_mod_pub_set(0, current_dev_addr, current_dev_addr,
                                          BT_MESH_MODEL_ID_HEALTH_SRV, &healthpub,
                                          &status));
        ASSERT_EQUAL(0, status);
 
-       k_sleep(K_SECONDS(1));
+       k_sleep(K_SECONDS(2));
 
        ASSERT_OK(bt_mesh_cfg_cli_node_reset(0, current_dev_addr, (bool *)&status));
```

cvinayak replied:
thanks will try it and get back on Monday

cvinayak committed:
tests: bsim: Bluetooth: mesh: Adjust test sleep durations

Adjust test sleep durations to compensate towards timing
changes in Zephyr Controller radio scheduling.

With the change in the radio scheduling, there is overlap
of advertising PDU from multiple nodes that relay the PDU,
hence increase test timeout so that next advertising
interval that uses random delay will get received by nodes
without collisions on air.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

carlescufi commented:
@mtpr-ot @wopu-ot could you please review?

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/ticker.h:
```diff
@@ -26,7 +26,19 @@
+ */
+#define HAL_TICKER_US_TO_TICKS_CEIL(x) \
+	( \
+		((uint32_t)(((((uint64_t) (x) * 1000000000UL) + \
+			      HAL_TICKER_CNTR_CLK_UNIT_FS - 1U)) / \
+			    HAL_TICKER_CNTR_CLK_UNIT_FS)) & \
+		HAL_TICKER_CNTR_MASK \
```
Would this be the same as 
```c
ceiling_fraction(((uint64_t) (x) * 1000000000UL), HAL_TICKER_CNTR_CLK_UNIT_FS) & HAL_TICKER_CNTR_MASK
```
?

cvinayak replied:
done

Thalley commented on tests/bluetooth/controller/mock_ctrl/include/hal/ticker_vendor_hal.h:
```diff
@@ -3,38 +3,57 @@
-	(((uint32_t)(((uint64_t)(x) * 1000000000UL) / 30517578125UL)) & HAL_TICKER_CNTR_MASK)
+#define HAL_TICKER_US_TO_TICKS(x) \
+	( \
+		((uint32_t)(((uint64_t) (x) * 1000000000UL) / \
```
The multiplication of `1000000000UL` to done several places in the controller code. Suggest to introduce a macro for, similar to the `NSEC_PER_MSEC` and similar macros

cvinayak replied:
done

Thalley commented on subsys/bluetooth/controller/ll_sw/openisa/hal/RV32M1/ticker.h:
```diff
@@ -35,6 +35,17 @@
+ */
+#define HAL_TICKER_US_TO_TICKS_CEIL(x) \
+	( \
+		((uint32_t)(((((uint64_t) (x) * 1000000000UL) + \
+			      HAL_TICKER_CNTR_CLK_UNIT_FS - 1U)) / \
+			    HAL_TICKER_CNTR_CLK_UNIT_FS)) & \
+		HAL_TICKER_CNTR_MASK \
```
Ditto with using `ceiling_fraction`

cvinayak replied:
done

Thalley commented on tests/bluetooth/controller/mock_ctrl/include/hal/ticker_vendor_hal.h:
```diff
@@ -29,6 +29,16 @@
+/* Macro to translate microseconds to tick units.
+ * NOTE: This returns the ceil value.
+ */
+#define HAL_TICKER_US_TO_TICKS_CEIL(x) \
+	( \
+		((uint32_t)(((((uint64_t) (x) * 1000000000UL) + \
+			      HAL_TICKER_CNTR_CLK_UNIT_FS - 1U)) / \
+			    HAL_TICKER_CNTR_CLK_UNIT_FS)) & \
+		HAL_TICKER_CNTR_MASK \
+	)
```
Which of the value(s) are vendor specific, since we need several versions of this macro? 

Could we introduce a generic macro and simply have the vendor specific value as input?

cvinayak replied:
The macro calculation is vendor specific. 
For platforms where one tick is one microsecond, then `HAL_TICKER_US_TO_TICKS_CEIL` would just be `(x)`. Hence, all the HAL_TICKER_* is currently vendor specific.

Thalley replied:
> For platforms where one tick is one microsecond, then HAL_TICKER_US_TO_TICKS_CEIL would just be (x). 

But in that could, couldn't we just do

```c
#define HAL_TICKER_US_TO_TICKS_CEIL(x, factor) \
	( \
		((uint32_t)(((((uint64_t) (x) * (factor)) + \
			      HAL_TICKER_CNTR_CLK_UNIT_FS - 1U)) / \
			    HAL_TICKER_CNTR_CLK_UNIT_FS)) & \
		HAL_TICKER_CNTR_MASK \
	)
```

cvinayak replied:
On nRF calculation is specific to have integer arithmetic when the clock is 32768 Hz, i.e. (1/32768)*1000000000 = 30517578125.

But say the clock is 16 MHz, then the calculation is simply ((x) * 16), it is cleaner to have separate hal/ticker.h file for different platforms with differing clock frequencies.

Thalley replied:
Fair point - No reason to change in this PR anyways :)

Thalley commented on tests/bsim/bluetooth/mesh/src/test_provision.c:
```diff
@@ -554,26 +554,26 @@ static void node_configure_and_reset(void)
 	ASSERT_EQUAL(0, status);
 
-	k_sleep(K_SECONDS(1));
+	k_sleep(K_SECONDS(2));
```
Does the changes to these test imply that scheduling is now "slower"? 

I think the commit message could use a bit more reasoning/elaboration.

cvinayak replied:
Updated the commit message, that the additional delay in timeout is introduced that so advertising retries are considered in verifying the test criteria.

cvinayak committed:
Bluetooth: Controller: Use defines for femto and pico seconds

Use defines for femto and pico seconds in HAL ticker
calculations.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Use DIV_ROUND_UP macro in HAL ticker

Use DIV_ROUND_UP macro in HAL ticker.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
@PavelVPV @Thalley Please re-review/approve, I had forgotten to fix `ticker_vendor_hal.h` in unit tests.
